## Miniprojeto - Análise de Dados com Pyhton T5 ##

In [1]:
%pip install -q seaborn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
# Importando as bibliotecas

import csv
from datetime import datetime
from IPython.display import display
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [3]:
# Realizando a leitura do CSV com DictReader (Manipulação de Arquivos CSV)

with open ("basevarejo.csv", encoding="utf-8") as arquivo:
    leitor = csv.DictReader(arquivo, delimiter=";")
    dados = list(leitor)

In [4]:
# Verificando as informações sobre a composição do arquivo

print("Quantidade de registros:", len(dados))
print("Colunas:", dados[0].keys())

dados[0]

Quantidade de registros: 830000
Colunas: dict_keys(['DATA', 'CO_ID', 'CL_ID', 'CL_GENERO', 'CL_EC', 'CL_FHL', 'CL_SEG', 'PR_ID', 'PR_CAT', 'PR_NOME', ''])


{'DATA': '01/02/2019',
 'CO_ID': '1000',
 'CL_ID': '534',
 'CL_GENERO': 'M',
 'CL_EC': '4',
 'CL_FHL': '1',
 'CL_SEG': 'C',
 'PR_ID': '67',
 'PR_CAT': 'BEBIDAS',
 'PR_NOME': 'REFRIGERANTE GUARANA',
 '': ''}

In [5]:
# Após verificação inicial dos dados, optei por converter para dataframe para facilitar e aglizar a análise exploratória

df = pd.DataFrame(dados)

In [6]:
# Validando as informações e dados do dataframe

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 830000 entries, 0 to 829999
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   DATA       830000 non-null  str  
 1   CO_ID      830000 non-null  str  
 2   CL_ID      830000 non-null  str  
 3   CL_GENERO  830000 non-null  str  
 4   CL_EC      830000 non-null  str  
 5   CL_FHL     830000 non-null  str  
 6   CL_SEG     830000 non-null  str  
 7   PR_ID      830000 non-null  str  
 8   PR_CAT     830000 non-null  str  
 9   PR_NOME    830000 non-null  str  
 10             830000 non-null  str  
dtypes: str(11)
memory usage: 69.7 MB


In [7]:
# Padronizando os nomes das colunas

df.columns = df.columns.str.strip().str.lower()

df.columns

Index(['data', 'co_id', 'cl_id', 'cl_genero', 'cl_ec', 'cl_fhl', 'cl_seg',
       'pr_id', 'pr_cat', 'pr_nome', ''],
      dtype='str')

In [8]:
# Verificando valores nulos ou ausentes

print(df.isnull().sum())

print((df == "").sum())

display(df["pr_nome"].value_counts())

display(df["pr_cat"].value_counts())

display(((df["pr_cat"] == "#N/D") & (df["pr_nome"] == "#N/D")).sum())

data         0
co_id        0
cl_id        0
cl_genero    0
cl_ec        0
cl_fhl       0
cl_seg       0
pr_id        0
pr_cat       0
pr_nome      0
             0
dtype: int64
data              0
co_id             0
cl_id             0
cl_genero         0
cl_ec             0
cl_fhl            0
cl_seg            0
pr_id             0
pr_cat            0
pr_nome           0
             830000
dtype: int64


pr_nome
PRESUNTO COZIDO          14381
SARDINHA                  7490
GEL                       7399
BANANA                    7385
DESENGORDURANTE           7378
                         ...  
#N/D                      3650
ABSORVENTE                3639
ALCOOL                    3611
ALIMENTO PARA PASSARO     3605
ALHO                      3574
Name: count, Length: 118, dtype: int64

pr_cat
ALIMENTOS     434767
HIGIENE       155574
LIMPEZA       145754
BEBIDAS        43299
PET            32399
ACESSORIOS     14557
#N/D            3650
Name: count, dtype: int64

np.int64(3650)

In [9]:
# Verificando a quantidade de registros duplicados

display(df.duplicated().sum())


np.int64(96553)

In [10]:
# Analisando os registros das duplicatas

duplicados = df[df.duplicated(keep=False)]

display(duplicados.head(20))

,data,co_id,cl_id,cl_genero,cl_ec,cl_fhl,cl_seg,pr_id,pr_cat,pr_nome,
3,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI,
7,01/02/2019,1000,534,M,4,1,C,11,ALIMENTOS,AZEITE,
14,01/02/2019,1000,534,M,4,1,C,13,ALIMENTOS,BANANA,
15,01/02/2019,1000,534,M,4,1,C,218,ALIMENTOS,BIFE DE COXAO MOLE,
19,01/02/2019,1000,534,M,4,1,C,13,ALIMENTOS,BANANA,
22,01/02/2019,1000,534,M,4,1,C,69,BEBIDAS,REFRIGERANTE LIMaO,
34,01/02/2019,1000,534,M,4,1,C,225,ALIMENTOS,ATUM,
40,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI,
46,01/02/2019,1000,534,M,4,1,C,11,ALIMENTOS,AZEITE,
49,01/02/2019,1000,534,M,4,1,C,69,BEBIDAS,REFRIGERANTE LIMaO,


In [11]:
# Como foram identificados itens repetidos na mesma compra, verificarei os produtos registrados em uma mesma compra
display(df[df['co_id'] == '1000'][['co_id', 'pr_id', 'pr_nome']])

,co_id,pr_id,pr_nome
0,1000,67,REFRIGERANTE GUARANA
1,1000,70,REFRIGERANTE OUTROS
2,1000,178,LENCO UMEDECIDO
3,1000,4,ABACAXI
4,1000,175,LIMPADOR MULTIUSO
5,1000,187,HASTES FLEXIVEIS
6,1000,163,MORTADELA
7,1000,11,AZEITE
8,1000,95,AMACIANTE
9,1000,198,ENERGETICO


Como não temos na base uma coluna sobre a quantidade de itens, optei por manter os registros duplicados, pois podem representar mais de uma unidade adquirida na compra.

In [12]:
# Conforme documentação da base, verificando a quantidade de compras e clientes únicos

print("Quantidade de compras:", df["co_id"].nunique())
print("Quantidade de clientes conforme cl_id:", df["cl_id"].nunique())

Quantidade de compras: 18471
Quantidade de clientes conforme cl_id: 1000


In [13]:
# Como a quantidade de compras não coincidiu com o informado no documento base (50.000), seguirei com análise possiveis inconsistências na coluna CO_ID

display(df["co_id"].value_counts())

display(df["co_id"].str.len().value_counts())

display ((~df["co_id"].str.isdigit()).sum())

datas_por_compra = df.groupby("co_id")["data"].nunique()

if (datas_por_compra > 1).any():
    print("Identificador de compra em mais de uma data!")
else:
    print("Cada identificador de compra possui uma única data!")

co_id
2224      89
3461      89
8022      89
27006     89
28006     89
          ..
908871     1
910819     1
913434     1
914348     1
915285     1
Name: count, Length: 18471, dtype: int64

co_id
6    741272
5     80802
4      7926
Name: count, dtype: int64

np.int64(0)

Cada identificador de compra possui uma única data!


In [14]:
# Removendo coluna vazia

df = df.drop(columns=[""])
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 830000 entries, 0 to 829999
Data columns (total 10 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   data       830000 non-null  str  
 1   co_id      830000 non-null  str  
 2   cl_id      830000 non-null  str  
 3   cl_genero  830000 non-null  str  
 4   cl_ec      830000 non-null  str  
 5   cl_fhl     830000 non-null  str  
 6   cl_seg     830000 non-null  str  
 7   pr_id      830000 non-null  str  
 8   pr_cat     830000 non-null  str  
 9   pr_nome    830000 non-null  str  
dtypes: str(10)
memory usage: 63.3 MB


In [15]:
# Verificando espaços extras nos valores das colunas de texto

for coluna in df.columns:
    if df[coluna].dtype == "str":
        quantidade = (df[coluna] != df[coluna].str.strip()).sum()
        print(coluna, ":", quantidade)

data : 0
co_id : 0
cl_id : 0
cl_genero : 0
cl_ec : 0
cl_fhl : 0
cl_seg : 0
pr_id : 0
pr_cat : 0
pr_nome : 0


In [16]:
# Substituindo na coluna pr_cat os valores #N/D por "Sem Categoria" e na coluna pr_nome os mesmos #N/D por "Não Informado"

categorias = []

for valor in df["pr_cat"]:
    if valor == "#N/D":
        categorias.append("Sem Categoria")
    else:
        categorias.append(valor)

df["pr_cat"] = categorias

display(df['pr_cat'].value_counts())


produtos = []

for valor in df["pr_nome"]:
    if valor == "#N/D":
        produtos.append("NÃO INFORMADO")
    else:
        produtos.append(valor)

df["pr_nome"] = produtos

display(df['pr_nome'].value_counts())

pr_cat
ALIMENTOS        434767
HIGIENE          155574
LIMPEZA          145754
BEBIDAS           43299
PET               32399
ACESSORIOS        14557
Sem Categoria      3650
Name: count, dtype: int64

pr_nome
PRESUNTO COZIDO          14381
SARDINHA                  7490
GEL                       7399
BANANA                    7385
DESENGORDURANTE           7378
                         ...  
NÃO INFORMADO             3650
ABSORVENTE                3639
ALCOOL                    3611
ALIMENTO PARA PASSARO     3605
ALHO                      3574
Name: count, Length: 118, dtype: int64

In [17]:
# Convertendo a coluna data com datetime

datas_convertidas = []

for data in df["data"]:
    datas_convertidas.append(datetime.strptime(data, "%d/%m/%Y"))

df["data"] = datas_convertidas

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 830000 entries, 0 to 829999
Data columns (total 10 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   data       830000 non-null  datetime64[us]
 1   co_id      830000 non-null  str           
 2   cl_id      830000 non-null  str           
 3   cl_genero  830000 non-null  str           
 4   cl_ec      830000 non-null  str           
 5   cl_fhl     830000 non-null  str           
 6   cl_seg     830000 non-null  str           
 7   pr_id      830000 non-null  str           
 8   pr_cat     830000 non-null  str           
 9   pr_nome    830000 non-null  str           
dtypes: datetime64[us](1), str(9)
memory usage: 63.3 MB


In [18]:
# Analisando a coluna da quantidade de filhos e realizando a conversão para int

print("Valores não numéricos:", (~df["cl_fhl"].str.isdigit()).sum())

df["cl_fhl"] = df["cl_fhl"].astype(int)

df.info()

df["cl_fhl"].value_counts()

Valores não numéricos: 0
<class 'pandas.DataFrame'>
RangeIndex: 830000 entries, 0 to 829999
Data columns (total 10 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   data       830000 non-null  datetime64[us]
 1   co_id      830000 non-null  str           
 2   cl_id      830000 non-null  str           
 3   cl_genero  830000 non-null  str           
 4   cl_ec      830000 non-null  str           
 5   cl_fhl     830000 non-null  int64         
 6   cl_seg     830000 non-null  str           
 7   pr_id      830000 non-null  str           
 8   pr_cat     830000 non-null  str           
 9   pr_nome    830000 non-null  str           
dtypes: datetime64[us](1), int64(1), str(8)
memory usage: 63.3 MB


cl_fhl
0    435501
2    106622
3    104685
1    102813
4     80379
Name: count, dtype: int64

In [19]:
# Verificando os valores das demais colunas referentes a informações sobre os clientes

display(df["cl_genero"].value_counts())
display(df["cl_ec"].value_counts())
display(df["cl_seg"].value_counts())

cl_genero
F    432576
M    397424
Name: count, dtype: int64

cl_ec
3    213742
4    202618
2    194990
1    194873
5     23777
Name: count, dtype: int64

cl_seg
B    530163
C    232101
A     67736
Name: count, dtype: int64

In [20]:
# Frequência de compras de acordo com o genêro dos clientes 

compras_genero = df.groupby("cl_genero")["co_id"].nunique()

clientes_genero = df.groupby("cl_genero")["cl_id"].nunique()

media_por_genero = (compras_genero / clientes_genero).round(2)

display(compras_genero)
display(clientes_genero)
display(media_por_genero)

cl_genero
F    9615
M    8856
Name: co_id, dtype: int64

cl_genero
F    519
M    481
Name: cl_id, dtype: int64

cl_genero
F    18.53
M    18.41
dtype: float64

## INSIGHT
Ao analisar a frequência de compras de acordo com o gênero, notei que as mulheres representam a maioria dos clientes e também possuem maior quantidade de compras. Ao considerar a média por cliente, as mulheres apresentam 18,53 compras e os homens 18,41, demonstrando uma diferença pequena na frequência de compras por gêneros. 

In [21]:
# Frequência de compras por estado civil

compras_ec = df.groupby("cl_ec")["co_id"].nunique()
clientes_ec = df.groupby("cl_ec")["cl_id"].nunique()
media_compras_ec = (compras_ec/clientes_ec).round(2)

display(compras_ec)
display(clientes_ec)
display(media_compras_ec)

cl_ec
1    4290
2    4377
3    4762
4    4551
5     491
Name: co_id, dtype: int64

cl_ec
1    233
2    238
3    250
4    251
5     28
Name: cl_id, dtype: int64

cl_ec
1    18.41
2    18.39
3    19.05
4    18.13
5    17.54
dtype: float64

In [22]:
# Criando uma base com um registro por cliente

clientes_estado_civil = df[['cl_id', 'cl_ec', 'cl_fhl']].drop_duplicates()

clientes_com_filhos = clientes_estado_civil[clientes_estado_civil["cl_fhl"]> 0]
display(clientes_com_filhos)

filhos_por_ec = clientes_com_filhos.groupby("cl_ec")["cl_id"].nunique()
display(filhos_por_ec)

percentual_filhos_ec = (filhos_por_ec/clientes_ec * 100).round(2)
display(percentual_filhos_ec)

ids_com_filhos = clientes_com_filhos["cl_id"]
compras_clientes_com_filhos = df[df["cl_id"].isin(ids_com_filhos)]

compras_filhos_ec = compras_clientes_com_filhos.groupby("cl_ec")["co_id"].nunique()

media_compras_filhos_ec = (compras_filhos_ec / filhos_por_ec).round(2)

display(compras_filhos_ec)
display(media_compras_filhos_ec)

,cl_id,cl_ec,cl_fhl
0,534,4,1
52,279,1,2
149,323,1,3
220,957,2,3
449,842,5,2
...,...,...,...
209155,930,4,1
214846,743,1,3
218760,854,5,1
236630,92,4,1


cl_ec
1    106
2    119
3    115
4    105
5     28
Name: cl_id, dtype: int64

cl_ec
1     45.49
2     50.00
3     46.00
4     41.83
5    100.00
Name: cl_id, dtype: float64

cl_ec
1    1917
2    2201
3    2210
4    1917
5     491
Name: co_id, dtype: int64

cl_ec
1    18.08
2    18.50
3    19.22
4    18.26
5    17.54
dtype: float64

## INSIGHT
Ao analisar o estado civil dos clientes que possuem filhos, notei que todos os clientes viúvos possuem filhos. Porém, ao analisar a frequência de compras, os clientes separados com filhos apresentam a maior média, com aproximadamente 19,22 compras por cliente, enquanto os viúvos apresentam a menor média, com 17,54.

In [23]:
# Calculando as estatísticas da quantidade de filhos por cliente

clientes = df[["cl_id", "cl_fhl"]].drop_duplicates()

print("Quantidade de clientes:", clientes["cl_id"].nunique())
print("Quantidade de registros:", len(clientes))

display(clientes["cl_fhl"].describe())

print("Moda:", clientes["cl_fhl"].mode())

Quantidade de clientes: 1000
Quantidade de registros: 1000


count    1000.00000
mean        1.13600
std         1.41333
min         0.00000
25%         0.00000
50%         0.00000
75%         2.00000
max         4.00000
Name: cl_fhl, dtype: float64

Moda: 0    0
Name: cl_fhl, dtype: int64


# INSIGHT
Ao analisar os dados, notei que apesar da média de filhos por cliente ser de 1,14, ao verificar o segundo quartil (50%) e a moda. fica perceptível que pelo menos
metade dos clientes não possui filhos, sendo 0 também a quantidade de filhos mais frequente na base.

In [24]:
# Analisando a quantidade de compras por segmento econômico e por cliente

compras_segmento = df.groupby("cl_seg")["co_id"].nunique()
display(compras_segmento)

clientes_segmento = df.groupby("cl_seg")["cl_id"].nunique()
display(clientes_segmento)

media_compras = (compras_segmento / clientes_segmento).round(2)
display(media_compras)

cl_seg
A     1492
B    11843
C     5136
Name: co_id, dtype: int64

cl_seg
A     84
B    645
C    271
Name: cl_id, dtype: int64

cl_seg
A    17.76
B    18.36
C    18.95
dtype: float64

## INSIGHT 
Após analisar a quantidade de por cliente de acordo com o segmento econômico, foi possível perceber que, apesar de o segmento B possuir maior número de clientes e compras,
o segmento C é o grupo que, segundo a média, realiza cmpras com maior frequência, apresentando aproximadamente 18,95 compras por cliente, enquanto os segmentos B e A 
apresentam médias de 18,36 e 17,76, respectivamente.

In [25]:
# Analisando as categorias presentes nas compras de cada segmento econômico

compras_categoria_segmento = df.groupby(["cl_seg", "pr_cat"])["co_id"].nunique()

display(compras_categoria_segmento)

segmento_a = compras_categoria_segmento["A"]
percentual_a = (segmento_a / compras_segmento["A"]) * 100
display(percentual_a.round(2))

segmento_b = compras_categoria_segmento["B"]
percentual_b = (segmento_b / compras_segmento["B"]) * 100
display(percentual_b.round(2))

segmento_c = compras_categoria_segmento["C"]
percentual_c = (segmento_c / compras_segmento["C"]) * 100
display(percentual_c.round(2))

cl_seg  pr_cat       
A       ACESSORIOS         726
        ALIMENTOS         1466
        BEBIDAS           1228
        HIGIENE           1430
        LIMPEZA           1403
        PET               1084
        Sem Categoria      261
B       ACESSORIOS        5852
        ALIMENTOS        11719
        BEBIDAS           9422
        HIGIENE          11283
        LIMPEZA          11200
        PET               8653
        Sem Categoria     2064
C       ACESSORIOS        2646
        ALIMENTOS         5082
        BEBIDAS           4085
        HIGIENE           4897
        LIMPEZA           4872
        PET               3758
        Sem Categoria      903
Name: co_id, dtype: int64

pr_cat
ACESSORIOS       48.66
ALIMENTOS        98.26
BEBIDAS          82.31
HIGIENE          95.84
LIMPEZA          94.03
PET              72.65
Sem Categoria    17.49
Name: co_id, dtype: float64

pr_cat
ACESSORIOS       49.41
ALIMENTOS        98.95
BEBIDAS          79.56
HIGIENE          95.27
LIMPEZA          94.57
PET              73.06
Sem Categoria    17.43
Name: co_id, dtype: float64

pr_cat
ACESSORIOS       51.52
ALIMENTOS        98.95
BEBIDAS          79.54
HIGIENE          95.35
LIMPEZA          94.86
PET              73.17
Sem Categoria    17.58
Name: co_id, dtype: float64

## INSIGHT
Analisando os dados de compras dos segmentos por categoria, é possível notar que todos são liderados por Alimentos, presente em mais de 98% das comprs, seguido por Higiene, com percentuais em torno de 95%, e Limpeza, com aproximadamente 94%. A ordem das categorias se mantém semelhante entre os segmentos, apresentando apenas algumas variações nos percentuais das demais categorias. Isso demonstra que apesar de pertecerem a segmentos econômicos diferentes, os clientes apresentam um padrão semelhante em relação às categorias presentes em suas compras.